cpp/ql/src/Likely Bugs/Likely Typos/illDefinedForLoop.ql

```diff
@@ -0,0 +1,102 @@
+/**
+ * @name Ill-defined for loop
+ * @description A for-loop iteration expressions goes backward with respect of the initialization statement and condition expression.
+ * @id cpp/ill-defined-for-loop
+ * @kind problem
+ * @problem.severity warning
+ * @precision high
+ * @tags external/microsoft/6293
+ * @msrc.severity important
+ */
+import cpp
+import semmle.code.cpp.rangeanalysis.SimpleRangeAnalysis
+
+predicate illDefinedDecrForStmt( ForStmt forstmt, Variable v, Expr initialCondition, Expr terminalCondition ) { 
+  v.getAnAssignedValue() = initialCondition 
```

Comment:
Something isn't quite right here.  `v.getAnAssignedValue()` gives you _every_ value assigned to `v`, not just the closest one  before the loop.  There are several options here, I think you could `import semmle.code.cpp.dataflow.DataFlow` and use:
```
DataFlow::localFlowStep(DataFlow::exprNode(initialCondition), DataFlow::exprNode(rel.getLesserOperand()))
```

Reply:
Good point. I didn't see any false positive, but it makes sense to fix it.

Reply:
I believe you'd see strange results on code such as this:
```
int i; // shared by both loops below
for (i = 10; i < 110; i--)
{
}
for (i = 20; i < 120; i--)
{
}
```